presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.predicate.ValueSet;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeUtils;
+import io.prestosql.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverByteSize;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxRowCount;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterSourceOperator
+        implements Operator
+{
+    public static class DynamicFilterSourceOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final List<Type> types;
+        private final List<Integer> filterChannels;
+        private final List<String> filterDynamicFilterNames;
+        private final DynamicFilterClientSupplier dynamicFilterClientSupplier;
+        private final String sourceId;
+        private final int expectedDrivers;
+
+        private boolean closed;
+        private int driverId;
+
+        public DynamicFilterSourceOperatorFactory(
+                int operatorId,
+                PlanNodeId planNodeId,
+                List<Type> types,
+                List<Integer> filterChannels,
+                List<String> filterDynamicFilterNames,
+                DynamicFilterClientSupplier dynamicFilterClientSupplier,
```

Comment:
May be it's worth abstracting the `DynamicFilterClientSupplier` into a `Consumer` so this class could work both with "short-circuit" dynamic filtering (in #931).

Reply:
Agreed, it would be good to use `DynamicFilterSourceOperator` in both places